plugins/com.google.cloud.tools.eclipse.googleapis/src/com/google/cloud/tools/eclipse/googleapis/internal/GoogleApiFactory.java

```diff
@@ -0,0 +1,185 @@
+import org.eclipse.core.net.proxy.IProxyChangeEvent;
+import org.eclipse.core.net.proxy.IProxyChangeListener;
+import org.eclipse.core.net.proxy.IProxyData;
+import org.eclipse.core.net.proxy.IProxyService;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * Class to obtain various Google cloud Platform related APIs.
+ */
+@Component
+public class GoogleApiFactory implements IGoogleApiFactory {
+
+  private static final int DEFAULT_TIMEOUT_MS = 1000;
+
+  private static final Logger logger = Logger.getLogger(GoogleApiFactory.class.getName());
+
+  private JsonFactory jsonFactory;
+  private Map<String, HttpTransport> transports = new HashMap<>();
+
+  private IProxyService proxyService;
+  private GoogleApiUrls googleApiUrls;
+
+  public GoogleApiFactory() {
+    this(new GoogleApiUrls());
+  }
+
+  @VisibleForTesting
+  GoogleApiFactory(GoogleApiUrls googleApiUrls) {
+    this.googleApiUrls = googleApiUrls;
+  }
+
+  @Activate
+  public void init() {
+    Preconditions.checkNotNull(proxyService, "proxyService is null");
+    jsonFactory = new JacksonFactory();
+    buildTransports();
+    proxyService.addProxyChangeListener(new IProxyChangeListener() {
+      @Override
+      public void proxyInfoChanged(IProxyChangeEvent event) {
+        buildTransports();
+      }
+    });
+  }
+
+  /**
+   * @return the CloudResourceManager/Projects API
+   */
+  @Override
+  public Projects newProjectsApi(Credential credential) {
+    synchronized (transports) {
+      checkFields();
+      CloudResourceManager resourceManager =
+          new CloudResourceManager.Builder(transports.get(googleApiUrls.cloudResourceManagerUrl()),
+                                           jsonFactory, credential)
+          .setApplicationName(CloudToolsInfo.USER_AGENT).build();
+      Projects projects = resourceManager.projects();
+      return projects;
+    }
+  }
+
+  /**
+   * @return the Appengine/Apps API
+   */
+  @Override
+  public Apps newAppsApi(Credential credential) {
+    synchronized (transports) {
+      checkFields();
+      Appengine appengine =
+          new Appengine.Builder(transports.get(googleApiUrls.appEngineAdminUrl()), jsonFactory,
+                                credential)
+          .setApplicationName(CloudToolsInfo.USER_AGENT).build();
+      Apps apps = appengine.apps();
+      return apps;
+    }
+  }
+
+  private void checkFields() {
+    Preconditions.checkNotNull(transports, "transport is null");
+    Preconditions.checkNotNull(jsonFactory, "jsonFactory is null");
+  }
+
+  private void buildTransports() {
+    synchronized (transports) {
+      try {
+        buildAppEngineAdminTransport();
+        buildCloudResourceManagerTransport();
+      } catch (URISyntaxException ex) {
+        logger.log(Level.SEVERE, "Could not create transport using the proxy settings", ex);
+        transports = null;
```

Comment:
I think this will cause NPE in many places onward. Even if you initialize it with an empty map, it will cause trouble in places where you do `transports.get(...)`. Probably it's better to leave `transport` as-is? At least it will keep previous, working transports.